nomos-services/key-management-system/src/keys/mod.rs

```diff
@@ -24,17 +26,27 @@ impl SecuredKey for Key {
     type PublicKey = PublicKeyEncoding;
     type Error = KeyError;
 
-    fn sign(&self, data: &Self::Payload) -> Result<Self::Signature, Self::Error> {
-        match (self, data) {
-            (Self::Ed25519(key), Self::Payload::Ed25519(data)) => {
-                key.sign(data).map(Self::Signature::Ed25519)
+    fn sign(&self, payload: &Self::Payload) -> Result<Self::Signature, Self::Error> {
+        match (self, payload) {
+            (Self::Ed25519(key), Self::Payload::Ed25519(payload)) => {
+                key.sign(payload).map(Self::Signature::Ed25519)
+            }
+            (Self::Ed25519(_), _) => {
+                Err(EncodingError::requires::<<Ed25519Key as SecuredKey>::Payload>().into())
```

Comment:
Mhh, this is once again a runtime error vs a compile-time error. I am thinking how we can change that? Ideally we should not be allowed to pass a wrong payload for a given key. What is the need to gather keys under an enum, instead of just implementing the `SecuredKey` for each key? If we want to prevent outside crates to implement the same, we can seal the trait so only we can implement it. But in this case I would need some help to understand why we need this, and whether we can get rid of it?

Reply:
Yeah no I am not arguing on either panic or errors, I am arguing whether we can avoid runtime problems by introducing compile-time guarantees. I can try to see if I find an interesting avenue as soon as I have some time to spare. But until then we can go ahead with this PR. I would prefer if the service panics for any internal invariant that is not upheld, while it returns an error if it's something the calling service should manage.

Reply:
The problem is that this invariant may come from calling the service with wrong values. So again, imo that should be a panic rather than an error, but we should do that in all places. I think we can manage to avoid with some macros (so we generate the code specific for the available types), but that would take time.